File: magazine.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include "magazine.h"
using namespace std;
int Magazine::magazine_count = 0;
// Конструктор без параметров
Magazine::Magazine() {
	publisher_name = "";
	title = "";
	frequency = "";
	page_count = 0;
	price = 0.0f;
	++magazine_count;
}
// Конструктор с 1 параметром
Magazine::Magazine(string title) {
	this->title = title;
	++magazine_count;
}
// Конструктор со всеми параметрами
Magazine::Magazine(string publisher_name, string title, string frequency, int page_count, float price) {
	Magazine::setMagazineData(publisher_name, title, frequency, page_count, price);
	++magazine_count;
}
// Сеттер
void Magazine::setMagazineData(string publisher_name, string title, string frequency, int page_count, float price) {
	this->publisher_name = publisher_name;
	this->title = title;
	this->frequency = frequency;
	this->page_count = page_count;
	this->price = price;
}
// Метод для пользовательского ввода данных
void Magazine::enterMagazineData() {
	cout << endl;
	cout << "Введите название издателя:" << endl;
	getline(cin, publisher_name);
	cout << "Введите название журнала:" << endl;
	getline(cin, title);
	cout << "Введите периодичность:" << endl;
	getline(cin, frequency);
	cout << "Введите кол-во страниц (не меньше 0):" << endl;
	while ((cin >> page_count).fail() || (0 > page_count)) {
		cin.clear();
		cout << "Ошибка. Введите кол-во страниц (не меньше 0):" << endl;
		cin.ignore();
	}
	cin.ignore();
	cout << "Введите цену (не меньше 0):" << endl;
	while ((cin >> price).fail() || (0.0f > price)) {
		cin.clear();
		cout << "Ошибка. Введите цену (не меньше 0):" << endl;
		cin.ignore();
	}
	cin.ignore();
	cout << endl;
	cout << "Журнал " << title << " был успешно создан!" << endl;
}
// Геттер
void Magazine::getMagazineData() {
	cout << endl;
	cout << "Информация о журнале:" << endl;
	cout << "Издатель: " << publisher_name << endl;
	cout << "Название журнала: " << title << endl;
	cout << "Периодичность: " << frequency << endl;
	cout << "Кол-во страниц: " << page_count << endl;
	cout << "Цена: " << price << endl;
}
// Метод для сравнения названий издателя
bool Magazine::isPublisherNamesSame(string publisher_name) {
	// Создание копий названий издателя для поиска (чтобы не портить их первоначальный вид ф-ей transform)
	string publisher_name_field_copy = this->publisher_name;
	string publisher_name_param_copy = publisher_name;
	// Приведение строк к одному регистру
	transform(publisher_name_field_copy.begin(), publisher_name_field_copy.end(), publisher_name_field_copy.begin(), tolower);
	transform(publisher_name_param_copy.begin(), publisher_name_param_copy.end(), publisher_name_param_copy.begin(), tolower);
	// Возврат результата сравнения
	return (publisher_name_field_copy == publisher_name_param_copy);
}
// Статический геттер для получения кол-ва объектов
int Magazine::getMagazineCount() {
	return magazine_count;
}
// Переопределение виртуального метода из базового абстрактного класса
void Magazine::read() {
	cout << "Журнал прочитан" << endl;
}
// Функция поиска журналов по издателю
void findMagazinesByPublisher(Magazine magazines[], int magazine_count, string publisher_name) {
	Magazine found_magazines[300];
	int j = 0;
	for (int i = 0; i < magazine_count; i++) {
		if (magazines[i].isPublisherNamesSame(publisher_name)) {
			found_magazines[j] = magazines[i];
			++j;
		}
	}
	if (j) {
		cout << endl;
		cout << "Найдено " << j << " журналов издателя " << publisher_name << ":" << endl;
		for (int k = 0; k < j; k++) {
			cout << endl;
			cout << "Журнал " << (k + 1) << endl;
			cout << endl;
			found_magazines[k].getMagazineData();
			cout << "-------------------------" << endl;
		}
	}
	else {
		cout << endl;
		cout << "Ничего не найдено!" << endl;
	}
}

File: src/com/github/Yazip/LR3/Magazine.java
package com.github.Yazip.LR3;
import java.util.Scanner;
// Класс журнал
public class Magazine extends Publication{
    private String publisher_name;
    private String title;
    private String frequency;
    private int page_count;
    private float price;
    private static int magazine_count = 0;
    // Конструктор без параметров
    public Magazine() {
        publisher_name = "";
        title = "";
        frequency = "";
        page_count = 0;
        price = 0.0f;
        ++magazine_count;
    }
    // Конструктор с 1 параметром
    public Magazine(String title) {
        this.title = title;
        ++magazine_count;
    }
    // Конструктор со всеми параметрами
    public Magazine(String publisher_name, String title, String frequency, int page_count, float price) {
        setMagazineData(publisher_name, title, frequency, page_count, price);
        ++magazine_count;
    }
    // Сеттер
    public void setMagazineData(String publisher_name, String title, String frequency, int page_count, float price) {
        this.publisher_name = publisher_name;
        this.title = title;
        this.frequency = frequency;
        this.page_count = page_count;
        this.price = price;
    }
    // Метод для пользовательского ввода данных
    public void enterMagazineData() {
        Scanner scanner = new Scanner(System.in, "UTF-8");
        System.out.println();
        System.out.println("Введите название издателя:");
        publisher_name = scanner.nextLine();
        System.out.println("Введите название журнала:");
        title = scanner.nextLine();
        System.out.println("Введите периодичность:");
        frequency = scanner.nextLine();
        System.out.println("Введите кол-во страниц (не меньше 0):");
        do {
            while (!scanner.hasNextInt()) {
                System.out.println("Ошибка. Введите корректное число:");
                scanner.next();
            }
            page_count = scanner.nextInt();
            if (page_count < 0) {
                System.out.println("Ошибка. Введите кол-во страниц не меньше 0:");
            }
        } while (page_count < 0);
        System.out.println("Введите цену (не меньше 0):");
        do {
            while (!scanner.hasNextFloat()) {
                System.out.println("Ошибка. Введите корректную цену:");
                scanner.next();
            }
            price = scanner.nextFloat();
            if (price < 0.0f) {
                System.out.println("Ошибка. Введите цену не меньше 0:");
            }
        } while (price < 0.0f);
        System.out.println();
        System.out.println("Журнал " + title + " был успешно создан!");
    }
    // Геттер
    public void getMagazineData() {
        System.out.println();
        System.out.println("Информация о журнале:");
        System.out.println("Издатель: " + publisher_name);
        System.out.println("Название журнала: " + title);
        System.out.println("Периодичность: " + frequency);
        System.out.println("Кол-во страниц: " + page_count);
        System.out.println("Цена: " + price);
    }
    // Метод для поиска журналов по издателю
    public static void findMagazinesByPublisher(Magazine[] magazines, String publisher_name) {
        Magazine[] found_magazines = new Magazine[300];
        int j = 0;
        for (Magazine magazine : magazines) {
            if ((magazine.publisher_name).equalsIgnoreCase(publisher_name)) {
                found_magazines[j] = magazine;
                ++j;
            }
        }
        if (j > 0) {
            System.out.println();
            System.out.println("Найдено " + j + " журналов издателя " + publisher_name + ":");
            for (int k = 0; k < j; k++) {
                System.out.println();
                System.out.println("Журнал " + (k + 1));
                System.out.println();
                found_magazines[k].getMagazineData();
                System.out.println("-------------------------");
            }
        }
        else {
            System.out.println();
            System.out.println("Ничего не найдено!");
        }
    }
    // Статический геттер для получения кол-ва объектов
    public static int getMagazineCount() {
        return magazine_count;
    }
    // Переопределение абстрактного метода из базового абстрактного класса
    @Override
    public void read() {
        System.out.println("Журнал прочитан");
    }
}
